File: baekjoon/4주차/6198-옥상-정원-꾸미기/정우/6198.py
import sys

N = int(sys.stdin.readline())
buildings = []
for _ in range(N):
    buildings.append(int(sys.stdin.readline()))
stack = [(buildings[-1], 0)]
cnt = 0
for i in range(N-2, -1, -1):
    t = s = 0
    while stack and stack[-1][0] < buildings[i]:
        s += stack.pop()[1]
        t += 1
    stack.append((buildings[i], t+s))
    cnt += t+s
print(cnt)

File: baekjoon/4주차/6198-옥상-정원-꾸미기/제문/6198.kt
package baekjoon

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val buildingNum = readLine().toInt()
    val buildings = LongArray(buildingNum)
    val stack = Stack<Long>()
    var answer = 0L

    for (i in 0 until buildingNum)
        buildings[i] = readLine().toLong()

    for (i in 0 until buildingNum) {
        while (stack.isNotEmpty() && stack.peek() <= buildings[i])
            stack.pop()

        answer += stack.size.toLong()
        stack.push(buildings[i])
    }

    print(answer)
}